File: hackerrank/problem solving/algorithms/2 - implementation/33 - library fine/library_fine.go
// https://www.hackerrank.com/challenges/library-fine/problem?isFullScreen=true

package main

import (
    "fmt"
    "time"
)

const (
    HackosDaysFine   int = 15
    HackosMonthsFine     = 500
    HackosYearsFine      = 10000
)

func main() {
    returnDate := readDate()
    dueDate := readDate()
    fmt.Println(calculateFine(returnDate, dueDate))
}

func readDate() time.Time {
    var day, month, year int
    _, _ = fmt.Scan(&day, &month, &year)
    return time.Date(year, time.Month(month), day, 0, 0, 0, 0, time.UTC)
}

// T: O(1)
// S: O(1) extra space
func calculateFine(returnDate, dueDate time.Time) int {
    if isYearLate(returnDate, dueDate) {
        return HackosYearsFine
    }
    if isMonthLate(returnDate, dueDate) {
        return (int(returnDate.Month()) - int(dueDate.Month())) * HackosMonthsFine
    }
    if isDayLate(returnDate, dueDate) {
        return (returnDate.Day() - dueDate.Day()) * HackosDaysFine
    }
    return 0
}

func isYearLate(returnDate, dueDate time.Time) bool {
    return returnDate.Year() > dueDate.Year()
}

func isMonthLate(returnDate, dueDate time.Time) bool {
    return returnDate.Year() == dueDate.Year() && returnDate.Month() > dueDate.Month()
}

func isDayLate(returnDate, dueDate time.Time) bool {
    return returnDate.Year() == dueDate.Year() && returnDate.Month() == dueDate.Month() && returnDate.Day() > dueDate.Day()
}

File: hackerrank/problem solving/algorithms/2 - implementation/33 - library fine/library_fine.c
// https://www.hackerrank.com/challenges/library-fine/problem?isFullScreen=true
// C23

#include <stdio.h>

enum hackos_fine {
    hackos_days_fine = 15,
    hackos_months_fine = 500,
    hackos_years_fine = 10000
};

typedef struct {
    int day, month, year;
} date;

date read_date();
int calculate_fine(const date *return_date, const date *due_date);
bool is_year_late(const date *return_date, const date *due_date);
bool is_month_late(const date *return_date, const date *due_date);
bool is_day_late(const date *return_date, const date *due_date);

int main()
{
    date return_date = read_date();
    date due_date = read_date();
    printf("%d\n", calculate_fine(&return_date, &due_date));
    return 0;
}

date read_date()
{
    date input_date;
    scanf("%d %d %d", &input_date.day, &input_date.month, &input_date.year);
    return input_date;
}

// T: O(1)
// S: O(1) extra space
int calculate_fine(const date *return_date, const date *due_date)
{
    if (is_year_late(return_date, due_date))
        return hackos_years_fine;
    if (is_month_late(return_date, due_date))
        return (return_date->month - due_date->month) * hackos_months_fine;
    if (is_day_late(return_date, due_date))
        return (return_date->day - due_date->day) * hackos_days_fine;
    return 0;
}

bool is_year_late(const date *return_date, const date *due_date)
{
    return return_date->year > due_date->year;
}

bool is_month_late(const date *return_date, const date *due_date)
{
    return return_date->year == due_date->year && return_date->month > due_date->month;
}

bool is_day_late(const date *return_date, const date *due_date)
{
    return return_date->year == due_date->year && return_date->month == due_date->month
           && return_date->day > due_date->day;
}
